fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/AggScalarSubQueryToWindowFunction.java

```diff
@@ -0,0 +1,371 @@
+                = ((LogicalFilter<LogicalProject<LogicalApply<Plan, LogicalAggregate<Plan>>>>) filter);
+        if (!check(node)) {
+            return filter;
+        }
+        return trans(node);
+    }
+
+    private boolean checkPattern(LogicalFilter filter) {
+        if (!(filter.child() instanceof LogicalProject)) {
+            return false;
+        }
+        LogicalProject project = (LogicalProject) filter.child();
+        if (project.child() == null || !(project.child() instanceof LogicalApply)) {
+            return false;
+        }
+        // filter(apply()) may be also ok.
+        // but nereids will match the pattern filter(project(apply()))
+        LogicalApply apply = ((LogicalApply<?, ?>) project.child());
+        if (!apply.isScalar() || !apply.isCorrelated() || !apply.getSubCorrespondingConject().isPresent()) {
+            return false;
+        }
+        return apply.left() != null && apply.right() instanceof LogicalAggregate
+                && ((LogicalAggregate<?>) apply.right()).child() != null;
+    }
+
+    private boolean check(LogicalFilter<LogicalProject<LogicalApply<Plan, LogicalAggregate<Plan>>>> node) {
+        LogicalApply<?, ?> apply = node.child().child();
+        LogicalPlan outer = ((LogicalPlan) apply.child(0));
+        LogicalPlan inner = ((LogicalPlan) apply.child(1));
+        outerPlans = new PlanCollector().collect(outer);
+        innerPlans = new PlanCollector().collect(inner);
+        LogicalFilter outerFilter = node;
+        Optional<LogicalFilter> innerFilter = innerPlans.stream()
+                .filter(LogicalFilter.class::isInstance)
+                .map(LogicalFilter.class::cast).findFirst();
+        return innerFilter.filter(
+                logicalFilter -> checkPlanType() && checkAggType()
+                        && checkRelation(apply.getCorrelationSlot())
+                        && checkPredicate(Sets.newHashSet(outerFilter.getConjuncts()),
+                        Sets.newHashSet(logicalFilter.getConjuncts()))
+        ).isPresent();
+    }
+
+    private Plan trans(LogicalFilter<LogicalProject<LogicalApply<Plan, LogicalAggregate<Plan>>>> node) {
+        LogicalApply<Plan, LogicalAggregate<Plan>> apply = node.child().child();
+        LogicalAggregate<Plan> agg = apply.right();
+
+        // for example, in tpc-h Q17
+        // window filter conjuncts is cast(l_quantity#id1 as decimal(27, 9)) < `0.2 * avg(l_quantity)`#id2 and
+        // 0.2 * avg(l_quantity#id3) as `0.2 * l_quantity`#id2 of agg's output expression
+        // we change it to cast(l_quantity#id1 as decimal(27, 9)) < 0.2 * `avg(l_quantity#id1) over(window)`#id4
+        // and avg(l_quantity#id1) over(window) as `avg(l_quantity#id1) over(window)`#id4
+        // first we get the slots of the two sides of apply
+
+        Preconditions.checkArgument(apply.getSubCorrespondingConject().get() instanceof ComparisonPredicate);
```

Comment:
why check here? where to ensure `apply.getSubCorrespondingConject()` is not empty and it  must be a `ComparisonPredicate`?
if no where to ensure this.
we should add this to check and return false if it is not satisfy the restriction